software/contrib/gates_and_triggers.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+"""Convert incoming triggers to gates or gates to triggers
+
+Also outputs a toggle input which changes state every time an incoming gate is received, or when either button is
+pressed.
+
+@author Chris Iverach-Brereton <ve4cib@gmail.com>
+@year   2023
+"""
+
+import time
+
+from europi import *
+from europi_script import EuroPiScript
+from math import sqrt
+
+from experimental.screensaver import OledWithScreensaver
+
+ssoled = OledWithScreensaver()
+
+## All digital output signals are 5V
+OUTPUT_VOLTAGE = 5.0
+
+## Trigger outputs are 10ms long (rising/falling edges of gate signals)
+TRIGGER_DURATION_MS = 10
+
+class GatesAndTriggers(EuroPiScript):
+    def __init__(self):
+        super().__init__()
+
+        now = time.ticks_ms()
+
+        self.on_incoming_rise_start_time = now
+        self.on_incoming_fall_start_time = now
+        self.on_gate_fall_start_time = now
+        self.on_toggle_fall_start_time = now
+
+        # assign each of the CV outputs to specific duties
+        self.gate_out = cv1
+        self.incoming_rise_out = cv2
+        self.incoming_fall_out = cv3
+        self.gate_fall_out = cv4
+        self.toggle_out = cv5
+        self.toggle_fall_out = cv6
+
+        self.gate_on = False
+        self.toggle_on = False
+
+        self.k1_percent = int(k1.percent() * 100)
+        self.k2_percent = int(k2.percent() * 100)
+
+        # turn off all outputs initially
+        for cv in cvs:
+            cv.voltage(0)
+
+        @din.handler
+        def on_din_rising():
+            self.on_rise()
```

Comment:
[discussion] The recommendation is to do as little as possible in the handler code. Typically just set a flag and then handle it in the main loop. If this is working well then there's no need to change anything prematurely, but if you start to run into performance issues (likely at higher bpm on `din`) then this is the first place I would look.

Reply:
So far I haven't seen any performance issues, but I've also been using it with pretty slow LFOs. Definitely something we could fix in a future patch if we get issue reports, but for now I'm not sure it's needed.